### PR TITLE
    ceph: set default FsGroupChangePolicy value to 'None'

### PR DESCRIPTION
**Pull request by humblec**
The None value Indicates that volumes will be mounted
    with no modifications, as the CSI volume driver does not
    support these operations it is safe to place with `none`. 
    While volumes are provisioned by the RBD CSI driver for
    filesystem/block mode PVCs we dont expect the Fsgroup
    policy or check from CO side to play a role here.
    
    The current default value ie ReadWriteOnceWithFSType cause
    volumes to be examined to determine if volume ownership and permissions
    should be modified to match the pod's security policy. Changes could occur
    if the fsType is defined and the persistent volume's accessModes contains
    ReadWriteOnce.
**Checklist:**

- [ ] **Commit Message Formatting**: Commit titles and messages follow guidelines in the [developer guide](https://rook.io/docs/rook/latest/development-flow.html#commit-structure).
- [ ] **Skip Tests for Docs**: Add the flag for skipping the build if this is only a documentation change. See [here](https://github.com/rook/rook/blob/master/INSTALL.md#skip-ci) for the flag.
- [ ] **Skip Unrelated Tests**: Add a flag to run tests for a specific storage provider. See [test options](https://github.com/rook/rook/blob/master/INSTALL.md#test-storage-provider).
- [ ] Reviewed the developer guide on [Submitting a Pull Request](https://rook.io/docs/rook/latest/development-flow.html#submitting-a-pull-request)
- [ ] Documentation has been updated, if necessary.
- [ ] Unit tests have been added, if necessary.
- [ ] Integration tests have been added, if necessary.
- [ ] Pending release notes updated with breaking and/or notable changes, if necessary.
- [ ] Upgrade from previous release is tested and upgrade user guide is updated, if necessary.
- [ ] Code generation (`make codegen`) has been run to update object specifications, if necessary.

### CODE REVIEW

**Comment by humblec** [comment]
Considering we have enough e2e tests here to make sure, everything is working as expected, this is ready for review.

Cc @travisn @leseb

**Comment by travisn** [comment]
To be clear, does the ceph-csi driver not support `ReadWriteOnceWithFSType`? 

If the ceph-csi driver does support `ReadWriteOnceWithFSType`, shouldn't modifying the permissions according to the pod's security policy be the right default?

**Comment by humblec** [comment]
> To be clear, does the ceph-csi driver not support `ReadWriteOnceWithFSType`?
> 
> If the ceph-csi driver does support `ReadWriteOnceWithFSType`, shouldn't modifying the permissions according to the pod's security policy be the right default?

The check is actually one layer above to ceph csi, also afaict, the fsgroup on the share is NOOP . However I can perform some more testing on this before we get this in.

**Review by leseb — APPROVED** [review]

**Comment by travisn** [comment]
> > To be clear, does the ceph-csi driver not support `ReadWriteOnceWithFSType`?
> > If the ceph-csi driver does support `ReadWriteOnceWithFSType`, shouldn't modifying the permissions according to the pod's security policy be the right default?
> 
> The check is actually one layer above to ceph csi, also afaict, the fsgroup on the share is NOOP . However I can perform some more testing on this before we get this in.

I'm still not clear, why do we need to change the default? What is broken with the `ReadWriteOnceWithFSType` default? 

Instead of changing the default, what if we just add a sentence that describes when it's recommended to change the default? For example, does this impact mount times of large volumes?

**Comment by humblec** [comment]
> > 
> > The check is actually one layer above to ceph csi, also afaict, the fsgroup on the share is NOOP . However I can perform some more testing on this before we get this in.
> 
> I'm still not clear, why do we need to change the default? What is broken with the `ReadWriteOnceWithFSType` default?

If we keep the default as `ReadWriteOnceWithFSTpe` and if its not giving any advantages rather only some disadvantages atleast in certain scenarios, we have to changes the default and that was the thought process behind this PR :)

> 
> Instead of changing the default, what if we just add a sentence that describes when it's recommended to change the default? For example, does this impact mount times of large volumes?

Yeah, we are seeing that, this default value delay the pod startup when there are many files in a volume. We are still waiting for the outcome after changing the value to `None`, additionally, I am also planning to carry some extra tests here. so , lets hold this PR till we have a feedback

**Review by travisn — CHANGES_REQUESTED** [review]
Holding while more testing is being completed.

**Comment by humblec** [comment]
> > Instead of changing the default, what if we just add a sentence that describes when it's recommended to change the default? For example, does this impact mount times of large volumes?
> 
> Yeah, we are seeing that, this default value delay the pod startup when there are many files in a volume. We are still waiting for the outcome after changing the value to `None`, additionally, I am also planning to carry some extra tests here. so , lets hold this PR till we have a feedback

@travisn we have update from a couple of setups that, if we keep the policy as `none`, the pod startup is very fast and they dont face the issue. so thats indeed good.  :+1:  ,  The only disadvantage with keeping `none` as the default value is that, in case the root vol permission get changed, it wont be validated/corrected by kubelet.  Thinking about the outcome or result of skipping the permission correction from kubelet is that, in worst case the application will fail to access/use the volume properly. If thats the situation, I expect the pod container could fail to start or work properly.  Can we consider it as a good sign for the `admin` to enable or keep the change policy to `OnRootMismatch` in their setup ? wdyt? Based on that, we could pick this PR into or keep the existing defaults.

**Comment by travisn** [comment]
> > > Instead of changing the default, what if we just add a sentence that describes when it's recommended to change the default? For example, does this impact mount times of large volumes?
> > 
> > 
> > Yeah, we are seeing that, this default value delay the pod startup when there are many files in a volume. We are still waiting for the outcome after changing the value to `None`, additionally, I am also planning to carry some extra tests here. so , lets hold this PR till we have a feedback
> 
> @travisn we have update from a couple of setups that, if we keep the policy as `none`, the pod startup is very fast and they dont face the issue. so thats indeed good. 👍 , The only disadvantage with keeping `none` as the default value is that, in case the root vol permission get changed, it wont be validated/corrected by kubelet. Thinking about the outcome or result of skipping the permission correction from kubelet is that, in worst case the application will fail to access/use the volume properly. If thats the situation, I expect the pod container could fail to start or work properly. Can we consider it as a good sign for the `admin` to enable or keep the change policy to `OnRootMismatch` in their setup ? wdyt? Based on that, we could pick this PR into or keep the existing defaults.

Is this understanding correct?
- With the current default `ReadWriteOnceWithFSType`, volumes with many files will be delayed at startup while permissions are set
- With the setting `none`, volumes with many files will be mounted quickly since permissions are not checked by the kubelet
- With the setting `none`, if permissions get changed unexpectedly, the app may not be able to access the volume and cause failures
- With the setting `OnRootMismatch`, the permissions will only be scanned if the root permissions are different than expected, so volumes should really only be updated when absolutely necessary. 

So your recommendation is to change the default to `OnRootMismatch`? That makes more sense to me than changing the default to `none` if my understanding is correct.

**Review by leseb — CHANGES_REQUESTED** [review]
I think changing to use `OnRootMismatch` makes more sense especially if FsGroup is used when mounting the volume.

File: pkg/operator/ceph/csi/spec.go [hunk 1]
```diff
@@ -527,13 +527,13 @@ func (r *ReconcileCSI) startDrivers(ver *version.Info, ownerInfo *k8sutil.OwnerI
...
 	if EnableRBD {
-		err = csiDriverobj.createCSIDriverInfo(r.opManagerContext, r.context.Clientset, RBDDriverName, k8sutil.GetValue(r.opConfig.Parameters, "CSI_RBD_FSGROUPPOLICY", string(k8scsi.ReadWriteOnceWithFSTypeFSGroupPolicy)))
+		err = csiDriverobj.createCSIDriverInfo(r.opManagerContext, r.context.Clientset, RBDDriverName, k8sutil.GetValue(r.opConfig.Parameters, "CSI_RBD_FSGROUPPOLICY", string(k8scsi.NoneFSGroupPolicy)))
 		if err != nil {
 			return errors.Wrapf(err, "failed to create CSI driver object for %q", RBDDriverName)
 		}
 	}
 	if EnableCephFS {
-		err = csiDriverobj.createCSIDriverInfo(r.opManagerContext, r.context.Clientset, CephFSDriverName, k8sutil.GetValue(r.opConfig.Parameters, "CSI_CEPHFS_FSGROUPPOLICY", string(k8scsi.ReadWriteOnceWithFSTypeFSGroupPolicy)))
+		err = csiDriverobj.createCSIDriverInfo(r.opManagerContext, r.context.Clientset, CephFSDriverName, k8sutil.GetValue(r.opConfig.Parameters, "CSI_CEPHFS_FSGROUPPOLICY", string(k8scsi.NoneFSGroupPolicy)))
```
Comment (Madhu-1) [inline pkg/operator/ceph/csi/spec.go:536]: as part of https://github.com/rook/rook/pull/8459 we are supposed to change the fsgroupPolicy to None but looks like as an overlook the default value never changed, the PR only changed the logging, and yaml files are changed :facepalm: @humblec what's the plan for cephfs fsgroup policy change? are we tracking both cephfs and rbd changes in this PR?

**Pull request closed without merging**